### PR TITLE
Initial add of torch.distributed.pipelining

### PR DESCRIPTION
**Pull request by kwen2501**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #124776

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

### CODE REVIEW

**Comment by awgu** [comment]
Random thought: Any chance that `torch.distributed.pipeline_parallel` is an option? :)

**Comment by kwen2501** [comment]
Thanks for the suggestion @awgu !
We did consider `torch.distributed.pipeline_parallel` but unfortunately it seems a bit long :) 
Also, some people may argue that pipelining is not a kind of parallelism in a strict sense.

**Comment by wanchaol** [comment]
> Thanks for the suggestion @awgu ! We did consider `torch.distributed.pipeline_parallel` but unfortunately it seemed a bit long :) Also, some people may argue that pipelining is not a kind of parallelism in a strict sense.

I was about to say I have a similar suggestion with @awgu, I feel `pipeline_parallel` aligns well with other parts of our offerings: fully_sharded_data_parallel, tensor_parallel.

> pipelining is not a kind of parallelism in a strict sense.

Curious why it's not a kind of parallelism?

**Comment by kwen2501** [comment]
> fully_sharded_data_parallel, tensor_parallel

Where are these names in our package offering?

> Curious why it's not a kind of parallelism?

It is really a subtle difference, this is an answer from Quora's bot:
> Pipelining and parallelism are both techniques used in computer architecture to improve performance, but they operate in different ways.
Pipelining involves breaking down the execution of instructions into a series of stages, where each stage performs a different part of the instruction. This allows multiple instructions to be processed simultaneously, with each stage working on a different instruction. As a result, the overall throughput of the processor is increased.
Parallelism, on the other hand, involves executing multiple instructions simultaneously by using multiple processing units. This can be achieved through techniques such as multi-core processors or multi-processor systems. Parallelism allows for true simultaneous execution of instructions, which can significantly improve overall system performance.
In summary, pipelining focuses on breaking down the execution of individual instructions into smaller stages to improve throughput, while parallelism involves executing multiple instructions at the same time using multiple processing units to improve overall system performance.

In short, pipelining focuses on breaking down a job, while parallelism focuses on having multiple workers do the same job.

**Comment by wconstab** [comment]
I think the name 'pipeline parallel' is sufficiently established that it doesn't matter what quora says, people know what it means and understand it. This is the obvious safe option to me.

Pipelining is arguably a correct name but also not one the ML community uses as often. It is also shorter and clean.  I would have been more convinced by this if I wasn't already so used to saying pp, but I am.

**Comment by kwen2501** [comment]
Thanks @wconstab. I agree that "pipeline parallel" is a well-known concept -- it is also what we use in our README. But, as a _package name_, I think it is too long. "PP" is short, but no descriptive enough. 

Ideally, I prefer a package name that's one word, such as "distributed", "compiler", "profiler", "export". It can also be two shortened meaningful words concatenated, such as "autograd". I think "pipepara" looks weird.

**Review by wconstab — APPROVED** [review]

File: torch/distributed/pipelining/README.md [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+
+# Run the pipeline with input `x`. Divide the batch into 4 micro-batches
+# and run them in parallel on the pipeline
+if rank == 0:
+    schedule.step(x)
+else:
+    output = schedule.step()
+```
+
+Note that since we split our model into three stages, we must run this script with three workers. For this example, we will use `torchrun` to run multiple processes within a single machine for demonstration purposes. We can collect up all of the code blocks above into a file named [example.py](https://github.com/pytorch/PiPPy/tree/main/examples/basic) and then run it with `torchrun` like so:
```
[the remark this answers is not in the document]
Comment (kwen2501) [reply torch/distributed/pipelining/README.md:174]: The temporary decision is leaving the examples there (not migrated). Eventually they would be hosted in some tutorial repo.
